Recompose the security overview and built-in authentication content to diataxis doc templates

michelle-purcell opened the pull request:
Recompose the existing QUARKUS SECURITY OVERVIEW [security-overview-concept.adoc] and BUILT-IN AUTHENTICATION SUPPORT [security-built-in-authentication-support-concept.adoc] guides into the diataxis framework .

The net...
- [x] Moved the `== Security architecture` section into a new file  on its own called `security-architecture-concept.adoc`.
- [x] Moved all sections relating to authentication mechanisms into a new file called `security-authentication-mechanisms.adoc`.
- [x] Moved `== Identity providers` into a new file called `security-identity-providers.adoc`.
- [x] Moved most of Built-in authentication support content into 2 new files:
  - security-proactive-authentication-concept.adoc
  - security-authentication-mechanisms.adoc.
- [x] Restructured the Security overview content and links.
- [x] Fixed existing links to the content that was moved.

Note: Identified style and grammar enhancements that also need to be fixed.  For now, this PR focuses on moving and restructuring as a priority. Will address the other content style issues in another iteration.

Ref: [QDOCS-86](https://issues.redhat.com/browse/QDOCS-86) & (some of) [QDOCS-88](https://issues.redhat.com/browse/QDOCS-88)

michelle-purcell committed:
Recompose security overview and built-in auth content to diataxis templates

Recomposing security overview content to diataxis

Recomposing security overview content to diataxis

Addressed Sergey's review comments

Further enhancements

fix typos grammar

Fixed broken links after recomposing to diataxis

Fixed headings

More link fixes

Fix mTLS case

Update docs/src/main/asciidoc/security-architecture-concept.adoc

Co-authored-by: Mickey Maler <mmaler@redhat.com>

Update docs/src/main/asciidoc/security-architecture-concept.adoc

Co-authored-by: Mickey Maler <mmaler@redhat.com>

Update docs/src/main/asciidoc/security-identity-providers-concept.adoc

Co-authored-by: Mickey Maler <mmaler@redhat.com>

Update docs/src/main/asciidoc/security-architecture-concept.adoc

Co-authored-by: Mickey Maler <mmaler@redhat.com>

Update docs/src/main/asciidoc/security-identity-providers-concept.adoc

Co-authored-by: Mickey Maler <mmaler@redhat.com>

Update docs/src/main/asciidoc/security-proactive-authentication-concept.adoc

Co-authored-by: Mickey Maler <mmaler@redhat.com>

Update docs/src/main/asciidoc/security-proactive-authentication-concept.adoc

Co-authored-by: Mickey Maler <mmaler@redhat.com>

Update docs/src/main/asciidoc/security-proactive-authentication-concept.adoc

Co-authored-by: Mickey Maler <mmaler@redhat.com>

Update docs/src/main/asciidoc/security-proactive-authentication-concept.adoc

Co-authored-by: Mickey Maler <mmaler@redhat.com>

Update docs/src/main/asciidoc/security-proactive-authentication-concept.adoc

Co-authored-by: Mickey Maler <mmaler@redhat.com>

Some style enhancements

Update docs/src/main/asciidoc/security-proactive-authentication-concept.adoc

Co-authored-by: Mickey Maler <mmaler@redhat.com>

Style enhancement: reduce wordiness

tweak

Fixed further comments

asciidocfix and added link

Fixed some typos

Fixed build error

2023-01-24T11:15:40.1095928Z [ERROR] asciidoctor: WARN: security-proactive-authentication-concept.adoc: line 86: section title out of sequence: expected level 1, got level 2

michelle-purcell commented on docs/src/main/asciidoc/security-overview-concept.adoc:
```diff
@@ -2,264 +2,33 @@
 Complete the steps in the xref:security-basic-authentication-tutorial.adoc[Secure a Quarkus application with Basic authentication] tutorial.
 After you have successfully secured your Quarkus application with basic HTTP authentication, you can increase the security further by adding more advanced authentication mechanisms, for example, OpenID Connect (OIDC) authentication.
 
-== Security architecture
```
Note: All of this H2 section of security-overview-concept.adoc is now moved as-is into a new concept topic in this PR named security-architecture-concept.adoc.

sberyozkin replied:
@michelle-purcell Good idea, having an architecture diagram in the dedicated concept doc would work well as well, I'll ping you later after the PR is done to discuss

michelle-purcell commented on docs/src/main/asciidoc/security-overview-concept.adoc:
```diff
@@ -2,264 +2,33 @@
-In other contexts, it is possible to have other parallel representations of the same information or parts of it, for example, `SecurityContext`
-for JAX-RS or `JsonWebToken` for JSON Web Tokens (JWT).
-
-== Authentication mechanisms
```
Note: All of this H2 section in security-overview-concept.adoc is now moved into a new concept doc file in this PR named security-authentication-mechanisms-concept.adoc.

michelle-purcell commented on docs/src/main/asciidoc/security-overview-concept.adoc:
```diff
@@ -2,264 +2,33 @@
-|No
-|===
-
-[[identity-providers]]
```
Note: All of this H2 section of security-overview-concept.adoc is now moved as-is into a new concept topic in this PR named security-identity-providers-concept.adoc.

michelle-purcell commented on docs/src/main/asciidoc/security-overview-concept.adoc:
```diff
@@ -2,264 +2,33 @@
 
 For more information about customizing Quarkus Security including reactive security, and how to register a security provider, see xref:security-customization.adoc[Security customization].
 
-== Combining authentication mechanisms
```
Note: All of this H2 section of security-overview-concept.adoc is now moved (as-is) to the end of a new concept topic in this PR named security-authentication-mechanisms-concept.adoc.

michelle-purcell commented on docs/src/main/asciidoc/security-overview-concept.adoc:
```diff
@@ -438,7 +193,3 @@ You can optionally apply a time limit to individual suppressions by adding an ex
 `<suppress until="2022-01-01Z">...</suppress>`
 You can adjust the expiry date if you need to.
 
-== Quarkus Security testing
```
This H2 section is moved further up within this overview topic. (Eventually this could evolve into a bigger how-to / tutorial.)

sberyozkin replied:
FYI, we already have a security-testing.adoc, this section is only meant to link to it

sberyozkin commented on docs/src/main/asciidoc/security-architecture-concept.adoc:
```diff
@@ -0,0 +1,38 @@
+
+In other contexts, it is possible to have other parallel representations of the same information or parts of it, for example, `SecurityContext` for JAX-RS or `JsonWebToken` for JSON Web Tokens (JWT).
+
+For more information, see xref:security-built-in-authentication-support.adoc[Built-in authentication support].
```
We should also link to the authentication mechanism and identity provider concepts.
Not sure about  the built in authentication link being present in this section - since it is mostly about providing info about the Form/Basic/MTLS authentication

michelle-purcell replied:
@sberyozkin - Thanks :+1: 

 + 1 to suggestion to update the link in line 23.  Will also add links to those other 2 new topics. 

I think that most of the content in `security-built-in-authentication-support-concept.adoc` will soon recomposed so I held off changing the link until that piece of the work is done. 

The linked built-in topic will probably be retired and its current content moved over into several other more focussed diataxis topics, including:

- security-authentication-mechanisms.adoc
- security-overview-concept.adoc
- security-architecture-concept.adoc
and possibly a new concept topic for Proactive authentication, as there is a lot of content to warrant a separate concept (and possibly a how-to).

What do you think?

@rolfedh - cc - we overlap here on the diataxis restructuring of `security-built-in-authentication-support-concept.adoc`.

sberyozkin replied:
@michelle-purcell 

>The linked built-in topic will probably be retired and its current content moved over into several other more focussed diataxis topics

Makes sense, `security-authentication-mechanisms` would be what `built-in auth`  is a concrete instance of, most of it . 

>and possibly a new concept topic for Proactive authentication, as there is a lot of content to warrant a separate concept

+1, I was just typing a similar suggestion and then noticed you had already thought about it :-)

michelle-purcell commented:
@sberyozkin - I have now pushed more commits to resolve the comments so far and complete the link updates that were required. This PR is now ready for final review. Thanks.

michelle-purcell commented:
@rolfedh , @inoxx03 , @sheilamjones , @MichalMaler, @ebullient, @any-other-community-sme-interested:  
FYI - This content restructuring PR is now ready for review.
(The focus here is on restructuring into diataxis. Additional styling and new content will be needed here but in the next iteration of changes.)

MichalMaler commented on docs/src/main/asciidoc/security-architecture-concept.adoc:
```diff
@@ -0,0 +1,42 @@
+* `IdentityProvider`
+* `SecurityidentityAugmentor`
+
+For more information about customizing Quarkus Security including reactive security and how to register a security provider, see xref:security-customization.adoc[Security customization].
```
```suggestion
For more information about customizing Quarkus Security, including reactive security and how to register a security provider, see xref:security-customization.adoc[Security customization].
```

MichalMaler commented on docs/src/main/asciidoc/security-identity-providers-concept.adoc:
```diff
@@ -0,0 +1,27 @@
+include::_attributes.adoc[]
+:categories: security
+In the Quarkus Security framework, identity providers play a key role in authentication and authorization, providing services for storing and verifying user identities.
+The JPA `IdentityProvider` creates a `SecurityIdentity` instance, which is used during user authentication to verify and authorize access requests making your Quarkus application secure.
```
```suggestion
The JPA `IdentityProvider` creates a `SecurityIdentity` instance used during user authentication to verify and authorize access requests, making your Quarkus application secure.
```

MichalMaler commented on docs/src/main/asciidoc/security-identity-providers-concept.adoc:
```diff
@@ -0,0 +1,27 @@
+
+If you use Basic or form-based authentication then you must add an `IdentityProvider` instance that can convert a username and password to a `SecurityIdentity` instance.
+
+To get started with security in Quarkus, we recommend that you first combine the Quarkus built-in Basic HTTP authentication with the JPA identity provider to enable role-based access control (RBAC).
```
```suggestion
To get started with security in Quarkus, we recommend combining the Quarkus built-in Basic HTTP authentication with the JPA identity provider to enable role-based access control (RBAC).
```

MichalMaler commented on docs/src/main/asciidoc/security-proactive-authentication-concept.adoc:
```diff
@@ -0,0 +1,151 @@
+
+Requests with an invalid credential will always be rejected, even when the page is public. 
+
+If you only want to authenticate when the target page requires authentication, you can change the default behavior.
```
```suggestion
You can change the default behavior if you only want to authenticate when the target page requires authentication.
```

MichalMaler commented on docs/src/main/asciidoc/security-proactive-authentication-concept.adoc:
```diff
@@ -0,0 +1,151 @@
+`quarkus.http.auth.proactive=false` 
+----
+
+If you disable proactive authentication then the authentication process will only be run when an identity is requested, either because there are security rules that require the user to be authenticated, or due to programmatic access to the current identity.
```
```suggestion
If you disable proactive authentication, then the authentication process will only be run when an identity is requested, either because security rules require the user to authenticate or due to programmatic access to the current identity.
```

MichalMaler commented on docs/src/main/asciidoc/security-proactive-authentication-concept.adoc:
```diff
@@ -0,0 +1,151 @@
+If you disable proactive authentication then the authentication process will only be run when an identity is requested, either because there are security rules that require the user to be authenticated, or due to programmatic access to the current identity.
+
+If proactive authentication is in use, accessing the `SecurityIdentity` is a blocking operation. 
+This is because authentication may not have happened yet, and accessing it may require calls to external systems such as databases that may block.
```
```suggestion
This is because authentication may have yet to happen, and accessing it may require calls to external systems, such as databases that may block.
```

MichalMaler commented on docs/src/main/asciidoc/security-proactive-authentication-concept.adoc:
```diff
@@ -0,0 +1,151 @@
+
+If proactive authentication is in use, accessing the `SecurityIdentity` is a blocking operation. 
+This is because authentication may not have happened yet, and accessing it may require calls to external systems such as databases that may block.
+For blocking applications this is no problem, however, if you have disabled authentication in a reactive application this will fail (as you cannot do blocking operations on the IO thread). 
```
```suggestion
For blocking applications, this is no problem. However, if you have disabled authentication in a reactive application, this will fail (as you cannot do blocking operations on the IO thread).
```

MichalMaler commented on docs/src/main/asciidoc/security-proactive-authentication-concept.adoc:
```diff
@@ -0,0 +1,151 @@
+If proactive authentication is in use, accessing the `SecurityIdentity` is a blocking operation. 
+This is because authentication may not have happened yet, and accessing it may require calls to external systems such as databases that may block.
+For blocking applications this is no problem, however, if you have disabled authentication in a reactive application this will fail (as you cannot do blocking operations on the IO thread). 
+To work around this you need to `@Inject` an instance of `io.quarkus.security.identity.CurrentIdentityAssociation`, and call the
```
```suggestion
To work around this, you need to `@Inject` an instance of `io.quarkus.security.identity.CurrentIdentityAssociation` and call the `Uni<SecurityIdentity> getDeferredIdentity();` method.
```

MichalMaler commented on docs/src/main/asciidoc/security-proactive-authentication-concept.adoc:
```diff
@@ -0,0 +1,151 @@
+This is because authentication may not have happened yet, and accessing it may require calls to external systems such as databases that may block.
+For blocking applications this is no problem, however, if you have disabled authentication in a reactive application this will fail (as you cannot do blocking operations on the IO thread). 
+To work around this you need to `@Inject` an instance of `io.quarkus.security.identity.CurrentIdentityAssociation`, and call the
+`Uni<SecurityIdentity> getDeferredIdentity();` method. 
```
```suggestion
```

MichalMaler commented on docs/src/main/asciidoc/security-proactive-authentication-concept.adoc:
```diff
@@ -0,0 +1,151 @@
+}
+----
+
+CAUTION: Some HTTP authentication mechanisms need to handle authentication exceptions themselves in order to create a correct authentication challenge.
```
```suggestion
CAUTION: Some HTTP authentication mechanisms need to handle authentication exceptions themselves to create a correct authentication challenge.
```

MichalMaler commented on docs/src/main/asciidoc/security-proactive-authentication-concept.adoc:
```diff
@@ -0,0 +1,151 @@
+CAUTION: Some HTTP authentication mechanisms need to handle authentication exceptions themselves in order to create a correct authentication challenge.
+For example, `io.quarkus.oidc.runtime.CodeAuthenticationMechanism` which manages OpenId Connect authorization code flow authentication, needs to build a correct redirect URL, cookies, etc.
+For that reason, using custom exception mappers to customize authentication exceptions thrown by such mechanisms is not recommended.
+In such cases, a safer way to customize authentication exceptions is to make sure the proactive authentication is not disabled and use Vert.x HTTP route failure handlers, as events come to the handler with the correct response status and headers.
```
```suggestion
In such cases, a safer way to customize authentication exceptions is to ensure the proactive authentication is not disabled and use Vert.x HTTP route failure handlers, as events come to the handler with the correct response status and headers.
```

MichalMaler commented:
Hello @michelle-purcell ! Great job on this topic. I spot a tag for a review, so  I added a few optional suggestions. Feel free to use them if liked :)
My review is done by this. Cheers!

sberyozkin commented on docs/src/main/asciidoc/security-oidc-bearer-authentication-concept.adoc:
```diff
@@ -915,6 +915,6 @@ Note Quarkus `web-app` applications always require `quarkus.oidc.client-id` prop
 * xref:security-openid-connect-dev-services.adoc[Dev Services for Keycloak]
 * xref:security-jwt-build.adoc[Sign and encrypt JWT tokens with SmallRye JWT Build]
-* xref:security-overview-concept.adoc#oidc-jwt-oauth2-comparison[Summary of Quarkus OIDC, JWT and OAuth2 features]
+* xref:security-authentication-mechanisms.adoc#combining-authentication-mechanisms[Combining authentication mechanisms]
```
This file has links to the OIDC web authentication at the very top in the introduction

sberyozkin replied:
If you'd like please keep the new `Combining authhentication mechanisms` as well, but IMHO, a link to the comparison of the token verification options should remain on its own

sberyozkin commented on docs/src/main/asciidoc/security-openid-connect-web-authentication.adoc:
```diff
@@ -1457,5 +1457,5 @@ It is important that this error endpoint is a public resource to avoid the user
 * xref:security-openid-connect-client-reference.adoc[OpenID Connect and OAuth2 Client and Filters Reference Guide]
 * xref:security-openid-connect-dev-services.adoc[Dev Services for Keycloak]
-* xref:security-overview-concept.adoc#oidc-jwt-oauth2-comparison[Summary of Quarkus OIDC, JWT and OAuth2 features]
+* xref:security-authentication-mechanisms.adoc#combining-authentication-mechanisms[Combining authentication mechanisms]
```
Sane comments as in the bearer adoc

sberyozkin commented:
@michelle-purcell Very nicely and carefully done, thanks, I can't see any problems, we might want to tweak a few things once this PR gets merged and we have a better appreciation of how it all interlinks and easy to consume, but right now it looks quite perfect, thanks. I've only proposed some minor link related updates

michelle-purcell commented:
@sberyozkin - Thank you for your thorough review and feedback :100: .  I have fixed all open comments. There's just one open comment from @mmaler that I fixed above that might need your SME input.

rolfedh reviewed (approved):
LGTM

sberyozkin commented:
Hi @michelle-purcell thanks,  I only did a few comments,  thanks again for preparing this PR, the open but fixed comment looks fine. There is some doc bulid issue, please fix tomorrow and we will merge,  thanks

sberyozkin commented:
Will be ready to go once the build failure is fixed

michelle-purcell commented:
@sberyozkin - All conflicts are now resolved. I think we are ready to merge. Thanks :+1:

sberyozkin commented:
Hi @michelle-purcell you might've used the wrong sha to rebase against, https://github.com/quarkusio/quarkus/pull/30422/commits/3c1d04ce428b7a1553f29e2719d771ef516526a2 is now in the list of commits.
@gsmet Hi Guillaume, can you please advise how to fix it ?

sberyozkin commented:
Cool, thanks for fixing it so fast :-)

michelle-purcell commented:
@sberyozkin & @gsmet - by the time I fixed my asciidoc source doc build issue, my branch was out of date! I've since rebased and all looks good now. Can we merge now? Thank you!

michelle-purcell commented:
@sberyozkin - FYI, added the 2.16 backport label to this merged security docs PR also.

gsmet commented:
I won't backport this one because of too many conflicts.

I think we need to be a lot more conservative about what we backport for documentation. Typos and small changes are OK but definitely not large structural changes.
